Updated Indexer FIP (ready to merge) (#365)
filecoin-project/FIPs: FIPS/IndexerFIP.md

## Change Motivation

Primary reasons why Indexer is critical to the success of Filecoin network:

1. **Content Discovery and Retrieval**: 
Running Indexer is required for efficient content discovery and retrieval for Filecoin network, since Filecoin needs a CID to Storage Provider mapping to enable content retrieval. It is also an important step toward retrieval across IPFS and Filecoin, as Indexer will index both IPFS and Filecoin data and support efficient and universal retrieval across IPFS and Filecoin network as the end state.
2. **Better usage and growth of the network**: 
Making data more accessible to the network increases Filecoin data usage. This drives adoption of the network to onboard more data, resulting in more  data being discoverable and retrievable - it is a positive growth flywheel.

> ## Terminology

Reviewer: Nit: I would move this and "Parties" to level-3 headings under "Specification", so the level-2 headings remain as in the template.

# Advertisement signals availability of content to the indexer nodes in form of a chunked list of 
# multihashes, where to retrieve them from, and over protocol they are retrievable.
type Advertisement struct {
    # PreviousID is an optional link to the previous advertisement.
    PreviousID optional Link
    # Provider is the peer ID of the host that provides this advertisement.
    Provider String
    # Addresses is the list of multiaddrs as strings from which the advertised content is retrievable.
    Addresses [String]
    # Signature is the signature of this advertisement.
    Signature Bytes
    # Entries is a link to the chained EntryChunk instances that contain the multihashes advertised. 
    Entries Link
    # ContextID is the unique identifier for the collection of advertised multihashes.
    ContextID Bytes
    # Metadata captures contextual information about how to retrieve the advertised content.
    Metadata Bytes
    # IsRm, when true, specifies that the content identified by ContextID is no longer retrievalbe from the provider.
    IsRm Bool
}
```

- The ContextID is limited to a maximum of 64 bytes.
- The Metadata is limited to a maximum of 1024 bytes (1KiB).

> The metadata field contains provider-specific data that pertains to the set of multihashes being added.  The metadata is prefixed with a protocol ID number followed by data encoded as per the protocol.  The content of the meta data is up to the provider, but if more then the limited size is needed, then the metadata should contain an ID identifying mode complex data stored by the provider.

Reviewer 2: Can we specify the filecoin-graphsync transport metadata format that is being used by filecoin here?

Author: bringing this to @gammazero 's attention

Reviewer 3: Hey @honghaoq - was @gammazero able to provide clarity here? Are any final changes needed?

Author: I am checking with Andrew

Reviewer 4: ```suggestion
The metadata field contains provider-specific data that pertains to the set of multihashes being added.  The metadata is prefixed with a protocol ID number followed by data encoded as per the protocol.  The content of the metadata is up to the provider, but if more than the limited size is needed, then the metadata should contain an ID identifying mode complex data stored by the provider.

Graphsync is the most commonly used protocol for retrieving content from a content provider. The filecoin-graphsync transport metadata is currently defined as follows:

Uvarint protcol `0x0910` (TransportGraphsyncFilecoinv1 in the [multicodec table](https://github.com/multiformats/multicodec/blob/master/table.csv#L134)). This is followed by a CBOR-encoded struct of:
- PieceCID, a link
- VerifiedDeal, boolean
- FastRetrieval, boolean  
```

---

> fip: "0034"

Reviewer 3: FIP0038 please.

title: Indexer Protocol for Filecoin Content Discovery
author: willscott, gammazero, honghaoq
status: Draft

> type: Technical (Core)

Reviewer 3: change type of 'FRC'

    ContextID Bytes
    # Metadata captures contextual information about how to retrieve the advertised content.
    Metadata Bytes
    # IsRm, when true, specifies that the content identified by ContextID is no longer retrievalbe from the provider.
    IsRm Bool
}
```

- The ContextID is limited to a maximum of 64 bytes.
- The Metadata is limited to a maximum of 1024 bytes (1KiB).

The metadata field contains provider-specific data that pertains to the set of multihashes being added.  The metadata is prefixed with a protocol ID number followed by data encoded as per the protocol.  The content of the metadata is up to the provider, but if more than the limited size is needed, then the metadata should contain an ID identifying mode complex data stored by the provider.

Graphsync is the most commonly used protocol for retrieving content from a content provider. The filecoin-graphsync transport metadata is currently defined as follows:

Uvarint protcol `0x0910` (TransportGraphsyncFilecoinv1 in the [multicodec table](https://github.com/multiformats/multicodec/blob/master/table.csv#L134)). This is followed by a CBOR-encoded struct of:

> - PieceCID, a link
> - VerifiedDeal, boolean
> - FastRetrieval, boolean  

Reviewer 4: ```suggestion
```go
type GraphsyncFilecoinV1 struct {
	// PieceCID identifies the piece this data can be found in
	PieceCID cid.Cid
	// VerifiedDeal indicates if the deal is verified
	VerifiedDeal bool
	// FastRetrieval indicates whether the provider claims there is an unsealed copy
	FastRetrieval bool
}
```
```

> # Indexer FIP

Reviewer 3: Hey @honghaoq! Can you please add in the correct header? [See FIP0001 for details. ](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0001.md#fip-header-preamble)

Author: Added!

---
fip: "0034"

> title: Indexer

Author: Updated to 'Indexer Protocol for Filecoin Content Discovery'

author: willscott, gammazero, honghaoq

> status: Final

Author: Updated!

type: Technical (Core)
created: 2022-04-26

> spec-pr: https://github.com/filecoin-project/FIPs/pull/365/files

Author: Added

- **Publisher**: Also referred to as _Index Provider_, This is an entity that publishes advertisements and index data to an indexer. It is usually, but not always, the same as the data provider. A publisher is identified by a libp2p peer ID.
- **Retrieval Addresses:** A list of addresses included in each advertisement that points to where you can retrieve from the advertised content.
- **Retrieval Client**: A client that queries an indexer to find where content is available, and retrieves that content from a provider.
- **Sync** (indexer with publisher): Operation that synchronizes the content indexed by an indexer with the content published by a publisher. A sync is initiated when an indexer receives and announces message, by an administrative command to sync with a publisher, or by the indexer when there have been no updates for a provider for some period of time (24 hours by default).

> ## Actors

Reviewer 5: ```suggestion
## Parties
```

> ## Actors

Author: Thanks, changed to 'Parties'

    {
	    "Multihash": "multihash-string",
	    "ProviderResults": [
        {
          "ContextID": "context-id-bytes",
          "Metadata": "metadata-bytes",
	        "Provider": {
            "ID": "peer-id-string",
            "Addrs": ["multiaddr-string", "multiaddr-string"]
          }
        }
      ]
    }
  ]
}
```

A Find result has a list of MultihashResults.  Each element of that list contains a Multihash and a list of ProviderResults for that multihash.  Each Provider result has a ContextID, Metadata, and Provider.  The Provider has an ID and a list of Addrs.

## **Backwards Compatibility**

This FIP does not change actors behavior so it does not require any Filecoin network update.

## **Test Cases**

The following test cases should be covered:

- Index Provider functionality test:
    - Ability to index storage deals
    - Storage Provider able to publish index to an Indexer node
- Scalability test: indexer need to be able to handle index storage at scale (hundreds of billions of index)
- Availability test:
    - 3-9s node uptime
    - Responsiveness to indexer advertisement queries
- Latency test:
    - Indexer query response latency: <300ms p95

>     - Indexer ingestion latency: <60 sec p95

Reviewer 5: ```suggestion
    - Indexing latency: <60 sec p95
```

>     - Indexer ingestion latency: <60 sec p95

Author: I added them as per the FIP template @kaitlin-beegle shared, happy to remove it though.

>     - Indexer ingestion latency: <60 sec p95

Reviewer: I think test cases are relevant to consensus-changing FIPs, but since this is a recommendation rather than requirement for node operators, they could have different performance.
